File: codigos/Campeonato.java
package codigos;

import java.util.*;
import java.util.stream.Collectors;

public class Campeonato {

    private final List<Partida> partidas = new ArrayList<>();
    private final Map<String, Time> times = new HashMap<>();

    // --------- Construtor ---------
    public Campeonato(
            List<Partida> partidas,
            Map<String, Time> times
    ) {
        this.partidas.addAll(partidas);
        this.times.putAll(times);

        // Atualiza as estatísticas de cada time baseado nas partidas
        for (Partida p : partidas) {
            Time mandante = this.times.get(p.getMandante());
            if (mandante != null) mandante.atualizarEstatisticas(p);

            Time visitante = this.times.get(p.getVisitante());
            if (visitante != null) visitante.atualizarEstatisticas(p);
        }
    }

    // --------- Métodos ---------
    public void adicionarPartida(Partida partida) {
        partidas.add(partida);

        Time mandante = times.get(partida.getMandante());
        if (mandante != null) mandante.atualizarEstatisticas(partida);

        Time visitante = times.get(partida.getVisitante());
        if (visitante != null) visitante.atualizarEstatisticas(partida);
    }

    public List<Partida> getPartidas() {
        return partidas;
    }

    public Map<String, Time> getTimes() {
        return times;
    }

    // --------- Classificações ---------
    public List<Time> getClassificacaoPorGols() {
        return times.values().stream()
                .sorted((a, b) -> b.getGolsPro() - a.getGolsPro())
                .collect(Collectors.toList());
    }

    public List<Time> getClassificacaoPorPontos() {
        return times.values().stream()
                .sorted((a, b) -> b.getPontuacao() - a.getPontuacao())
                .collect(Collectors.toList());
    }

    public List<Time> getClassificacaoPorSaldoGols() {
        return times.values().stream()
                .sorted((a, b) -> b.getSaldoGols() - a.getSaldoGols())
                .collect(Collectors.toList());
    }

    public List<Time> getClassificacaoPorVitoriasMandante() {
        return times.values().stream()
                .sorted((a, b) -> b.getVitoriasMandante() - a.getVitoriasMandante())
                .collect(Collectors.toList());
    }

    public List<Time> getClassificacaoPorVitoriasVisitante() {
        return times.values().stream()
                .sorted((a, b) -> b.getVitoriasVisitante() - a.getVitoriasVisitante())
                .collect(Collectors.toList());
    }

    // --------- Filtros ---------
    public List<Partida> getPartidasPorAno(int ano) {
        return partidas.stream()
                .filter(p -> {
                    String[] partes = p.getData().split("/");
                    int anoPartida = Integer.parseInt(partes[2]);
                    return anoPartida == ano;
                })
                .collect(Collectors.toList());
    }
}

File: Python/codigosPython/Campeonato.py
from typing import List, Dict
from .Partida import Partida
from .Time import Time

class Campeonato:
    def __init__(self, partidas: List[Partida], times: Dict[str, Time]):
        self._partidas: List[Partida] = list(partidas)
        self._times: Dict[str, Time] = dict(times)

        for p in partidas:
            self._atualizarEstatisticasTimes(p)


    def _atualizarEstatisticasTimes(self, partida: Partida):
        mandante = self._times.get(partida.mandante)
        if mandante:
            mandante.atualizarEstatisticas(partida)

        visitante = self._times.get(partida.visitante)
        if visitante:
            visitante.atualizarEstatisticas(partida)


    def adicionarPartida(self, partida: Partida):
        self._partidas.append(partida)
        self._atualizarEstatisticasTimes(partida)


    @property
    def partidas(self) -> List[Partida]:
        return self._partidas.copy() 

    @property
    def times(self) -> Dict[str, Time]:
        return dict(self._times)  

    def getClassificacaoPorGols(self) -> List[Time]:
        return sorted(self._times.values(), key=lambda t: t.golsPro, reverse=True)

    def getClassificacaoPorPontos(self) -> List[Time]:
        return sorted(self._times.values(), key=lambda t: t.getPontuacao(), reverse=True)

    def getClassificacaoPorSaldoGols(self) -> List[Time]:
        return sorted(self._times.values(), key=lambda t: t.getSaldoGols(), reverse=True)

    def getClassificacaoPorVitoriasMandante(self) -> List[Time]:
        return sorted(self._times.values(), key=lambda t: t.vitoriasMandante, reverse=True)

    def getClassificacaoPorVitoriasVisitante(self) -> List[Time]:
        return sorted(self._times.values(), key=lambda t: t.vitoriasVisitante, reverse=True)


    def getPartidasPorAno(self, ano: int) -> List[Partida]:
        partidas_ano = []
        for p in self._partidas:
            try:
                anoPartida = int(p.data.split("/")[2])
                if anoPartida == ano:
                    partidas_ano.append(p)
            except (IndexError, ValueError):
                continue  
        return partidas_ano
